Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -196,4 +209,143 @@ lemma two_mul_semilength_eq_length : 2 * p.semilength = p.toList.length := by
+
+include h in
+lemma firstReturn_pos : 0 < p.firstReturn := by
+  by_contra! f
+  rw [Nat.le_zero, firstReturn, findIdx_eq] at f
+  · simp only [get_eq_getElem, getElem_range] at f
+    rw [← p.cons_tail_dropLast_concat h] at f
+    simp at f
+  · rw [length_range, length_pos]
+    exact toList_ne_nil.mpr h
+
+include h in
+lemma firstReturn_lt_length : p.firstReturn < p.toList.length := by
+  have lp := length_pos_of_ne_nil (toList_ne_nil.mpr h)
+  rw [← length_range p.toList.length]
+  apply findIdx_lt_length_of_exists
+  simp only [mem_range, decide_eq_true_eq]
+  use p.toList.length - 1
+  exact ⟨by omega, by rw [Nat.sub_add_cancel lp, take_of_length_le (le_refl _),
+    p.count_U_eq_count_D]⟩
+
+include h in
+lemma count_take_firstReturn_add_one :
+    (p.toList.take (p.firstReturn + 1)).count U = (p.toList.take (p.firstReturn + 1)).count D := by
+  have := findIdx_get (w := (length_range p.toList.length).symm ▸ firstReturn_lt_length h)
+  simpa using this
+
+lemma count_lt_count_of_lt_firstReturn {i : ℕ} (hi : i < p.firstReturn) :
+    (p.toList.take (i + 1)).count D < (p.toList.take (i + 1)).count U := by
+  have ne := not_of_lt_findIdx hi
+  rw [decide_eq_true_eq, ← ne_eq, get_eq_getElem, getElem_range] at ne
+  exact lt_of_le_of_ne (p.count_D_le_count_U (i + 1)) ne.symm
+
+@[simp]
+lemma firstReturn_add : (p + q).firstReturn = if p = 0 then q.firstReturn else p.firstReturn := by
+  split_ifs with h; · simp [h]
+  have u : (p + q).toList = p.toList ++ q.toList := rfl
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    have v := firstReturn_lt_length h
+    constructor
+    · rw [take_append_eq_append_take, show p.firstReturn + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil, count_take_firstReturn_add_one h]
+    · intro j hj
+      rw [take_append_eq_append_take, show j + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil]
+      exact (count_lt_count_of_lt_firstReturn hj).ne'
+  · rw [length_range, u, length_append]
+    exact Nat.lt_add_right _ (firstReturn_lt_length h)
+
+@[simp]
+lemma firstReturn_nest : p.nest.firstReturn = p.toList.length + 1 := by
+  have u : p.nest.toList = U :: p.toList ++ [D] := rfl
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_of_length_le (by simp), ← u, p.nest.count_U_eq_count_D]
+    · intro j hj
+      simp_rw [cons_append, take_cons, count_cons, beq_self_eq_true, ite_true,
+        beq_iff_eq, ite_false, take_append_eq_append_take,
+        show j - p.toList.length = 0 by omega, take_zero, append_nil]
+      have := p.count_D_le_count_U j
+      omega
+  · simp_rw [length_range, u, length_append, length_cons]
+    exact Nat.lt_add_one _
+
+lemma nest_add_firstReturn : (p.nest + q).firstReturn = p.toList.length + 1 := by
+  rw [firstReturn_add, firstReturn_nest, ite_eq_right_iff, ← toList_eq_nil]; intro; contradiction
```

Comment:
```suggestion
lemma firstReturn_nest_add : (p.nest + q).firstReturn = p.toList.length + 1 := by simp
```
also aybe we don't need that lemma at all as it's basically `firstReturn_add` + `firstReturn_nest`? Can you make sure simp proves it?

Reply:
Okay, maybe you want this lemma because `p.nest + q` is a common idiom, but please make sure simp works